src/pygeoprocessing/routing/routing.pyx

```diff
@@ -850,34 +860,45 @@ def fill_pits(
                 # if it's a pit or plateau
                 search_queue.push(CoordinateType(xi_root, yi_root))
                 flat_region_mask_managed_raster.set(xi_root, yi_root, 1)
-                downhill_drain = 0
-                nodata_drain = 0
+                natural_drain_exists = 0
 
                 # this loop does a BFS starting at this pixel to all pixels
                 # of the same height. the _drain variables are used to
                 # remember if a drain was encountered. it is preferable to
                 # search the whole region even if a drain is encountered, so
                 # it can be entirely marked as processed and not re-accessed
                 # on later iterations
+                search_steps = 0
                 while not search_queue.empty():
                     xi_q = search_queue.front().xi
                     yi_q = search_queue.front().yi
                     search_queue.pop()
+                    if search_steps > max_pixel_fill_count:
+                        # clear the search queue and quit
+                        while not search_queue.empty():
+                            search_queue.pop()
+                        natural_drain_exists = 1

```

Comment:
I get the use of `natural_drain_exists` over multiple variable indicators. But in this case, if we exit because of a "natural depression", then `natural_drain_exists` set to `true` is not really the case, it doesn't naturally drain! I'm not suggesting an edit or adding more variables, but this is a thought that jumped out to me.

Reply:
I was thinking that if there's a natural depression then that is the natural drain for the region, such as a lake or culverin. That's what I'm trying to capture here - not everything drains to the ocean. I get how it sounds confusing though. I've been trying to think of some alternatives... At first I was thinking "sink"? But that's really not a drain for the places that do drain. And this bit of code comes out of a case where I *did* have two variables checking the same thing `drains_to_nodata` `drains_to_edge` and it didn't matter, hence the change to `natural_drain_exists`. To me this seems fine, but if you've got an idea that's clearer I'm happy to use it.

Reply:
Nope, that sounds good!